src/Parallel/Main.hpp

```diff
@@ -671,6 +767,62 @@ void contribute_to_phase_change_reduction(
         "reduction.");
   }
 }
+
+template <typename Metavariables>
+std::tuple<std::string, std::string, size_t>
+Main<Metavariables>::checkpoint_dir_basename_pad() const noexcept {
+  const std::string basename = "SpectreCheckpoint";
+  constexpr size_t pad = 6;
+
+  const std::string counter = std::to_string(checkpoint_dir_counter_);
+  const std::string padded_counter =
+      std::string(pad - counter.size(), '0').append(counter);
+  const std::string checkpoint_dir = basename + padded_counter;
+  return std::make_tuple(checkpoint_dir, basename, pad);
+};
+
+template <typename Metavariables>
+std::string Main<Metavariables>::checkpoint_dir() const noexcept {
+  const auto [checkpoint_dir, basename, pad] = checkpoint_dir_basename_pad();
+  (void)basename;
+  (void)pad;
+  if (file_system::check_if_dir_exists(checkpoint_dir)) {
+    ERROR("Can't write checkpoint: dir " + checkpoint_dir + " already exists!");
+  }
+  return checkpoint_dir;
+};
+
+template <typename Metavariables>
+void Main<Metavariables>::check_future_checkpoint_dirs_available()
+    const noexcept {
+  // Can't lambda-capture from structured binding in clang-11
+  std::string checkpoint_dir;
+  std::string basename;
+  size_t pad;
+  std::tie(checkpoint_dir, basename, pad) = checkpoint_dir_basename_pad();
+
+  // Find existing files with names that match the checkpoint dir name pattern
+  const auto all_files = file_system::ls();
+  const std::regex re(basename + "[0-9]{" + std::to_string(pad) + "}");
+  std::vector<std::string> matching_files;
```

Comment:
maybe name this `checkpoint_files` or something indicating this is the list of files/dirs that match the checkpoint name?